### ISSUE TITLE
[Feature Request] - Expose zlib strategy '4' for fixed Huffman blocks.

### ISSUE THREAD

**Issue by ace-dent** [shssoichiro/oxipng#420]
As with OptiPNG, the zlib strategy `zs` may only be specified for 0-3, where:
0. Z_DEFAULT_STRATEGY
1. Z_FILTERED
2. Z_HUFFMAN_ONLY
3. Z_RLE
4. Z_FIXED

Using `zs4` might be useful for some small images, where static /fixed Huffman Tables are used.
It should be available at least for experimentation.

**Comment by ace-dent** [comment]
From further testing with OptiPNG v0.7.7, it seems zlib strategies `0` and `1` both can produce files with static Huffman blocks. I'm not sure what is happening internally in zlib with those strategy modes to enable this.

**Comment by andrews05** [comment]
Since we libdeflate/zopfli now and not zlib, this is no longer applicable so I think we can close 🙂

**Comment by TPS** [comment]
> Using `zs4` might be useful for some small images, where static /fixed Huffman Tables are used.
> It should be available at least for experimentation.

@andrews05 Are static/fixed Huffman tables available in either `libdeflate` &/or `zopfli`, & available to experiment w/ in `oxipng`?  Else, maybe keep open & update title?

**Comment by andrews05** [comment]
@TPS Neither of them expose any options for dealing with huffman, they handle it all automatically.

**Comment by TPS** [comment]
I wonder if either of them even try static tables. 🤔

**Comment by TPS** [comment]
I wonder if either of them even _try_ static tables. 🤔

**Comment by AlexTMjugador** [comment]
Zopfli tried both static and dynamic Huffman codes and automatically chooses what yields the smallest output.

**Comment by ace-dent** [comment]
@TPS & @AlexTMjugador - Just had confirmation from an author of Zopfli: It does **not** try static tables; only dynamic.
Surprisingly, this mode can be quite useful for certain small images.

**Comment by AlexTMjugador** [comment]
Then I'm not sure what's going on here then, as the `zopfli` crate, which is closely based on the original Zopfli implementation, [has code that is used to try and use different types of blocks](https://github.com/zopfli-rs/zopfli/blob/109d21174e637b8c0a2167d67bc381c739e6ecab/src/deflate.rs#L1082-L1129).

**Comment by ace-dent** [comment]
> I'm not sure what's going on

-- Me neither! The conversation with Jyrki Alakuijala is on the `JXL` Discord (https://discord.com/channels/794206087879852103/806898911091753051/1128612477609984131)

**Comment by TPS** [comment]
Is it possible that `zopfli-rs` added it independently or from a fork of the original `zopfli`?  Or does that code section refer to something else?

**Comment by andrews05** [comment]
As best I can tell from looking at the code, it does technically support fixed via the lower level ZopfliDeflate function. The higher level functions though, including ZopfliCompress and ZopfliZlibCompress (and therefore the cli tools), only ever use dynamic.

There’s also this great comment in `squeeze.h`: “The fixed tree never gives the best compression.” 😂
Note that while this would not have been true under zlib, zopfli tries much harder to form an optimal dynamic tree than zlib ever did, so it’s possible it is actually true here.

**Comment by TPS** [comment]
> Using `zs4` might be useful for some small images, where _static /fixed Huffman Tables_ are used.
> It should be available at least for experimentation.

> There’s also this great comment in `squeeze.h`: “The fixed tree never gives the best compression.” 😂

I think these might not be mutually exclusive, but I might be assuming an important detail which I should confirm:
The theoretical potential of the fixed table is that it isn't stored w/in the image, since it's known, anyway & so doesn't take up that many bytes of the image size, right?

Is there _any **known**_ example of such an image, that isn't better compressed dynamically for an overall smaller size, that's purely due solely to this?

**Comment by andrews05** [comment]
I believe you are correct there. Checking libdeflate, it actually does calculate the best (smallest) block type to use when outputting each block. So it's just zopfli that _hypothetically_ might give less optimal output on _extremely small_ data. But the difference would be around a couple of bytes.

**Comment by AlexTMjugador** [comment]
> Is it possible that `zopfli-rs` added it independently or from a fork of the original `zopfli`?

I think that may be the case with Rust's `zopfli`. To prove it, I did a quick test with a `random_data` file generated with `dd if=/dev/random bs=1 count=128 > random_data` (i.e., a command that outputs 128 bytes of random data), and compressed it with [`zopfli`'s CLI](https://github.com/zopfli-rs/zopfli/blob/main/src/main.rs), which uses the same compression code as OxiPNG. [`infgen`](https://github.com/madler/infgen) revealed that the generated GZip file contained a single uncompressed DEFLATE block:

```
! infgen 3.0 output
!
gzip
!
last
stored
data 246 216 218 '6Y
data 27 'Kn
data 225 'i
data 29 '9
data 205 128 142 'Q
data 228 149 'mT
data 9 '*
data 163 '<
data 135 'w
data 246 194 175 'A
data 231 'l
data 152 'm
data 181 204 'b|
data 130 '=
data 0 '(
data 20 'wB
data 4 'W@
data 127 '2"
data 163 'z
data 158 172 202 155 0 '4VS5
data 7 '&)
data 31 14 'a
data 157 211 'B
data 214 243 '*!
data 174 8 ']
data 26 175 165 ':
data 226 'q
data 213 '}
data 229 198 155 184 245 169 250 221 140 'd
data 235 14 168 '$W
data 212 '#
data 22 '0
data 246 '_ZW
data 19 228 'n7|l{
data 226 '%
data 225 183 20 '2
data 252 '3
data 209 '=
data 7 167
end
!
crc
length
```

Therefore, the hypothesis that Rust's `zopfli` crate only generates DEFLATE blocks with dynamic Huffman codes is false.

**Comment by andrews05** [comment]
Oh sorry, I got mixed up there - my findings in the C version don't apply to oxipng 🫢

That said, I looked a little closer and found I was wrong anyway 🫢🫢
There are more comments confirming the C version does do Fixed:
```
If btype=2 is specified, it tries all block types.
```
```
Whether to perform the expensive calculation of creating an optimal block
with fixed huffman tree to check if smaller. Only do this for small blocks or
blocks which already are pretty good with fixed huffman tree.
```

**Comment by TPS** [comment]
> > Is it possible that `zopfli-rs` added it independently or from a fork of the original `zopfli`?
> 
> I think that may be the case with Rust's `zopfli`. To prove it, I did a quick test

> ```
> ! infgen 3.0 output
> !
> gzip
> !
> last
> stored

> Therefore, the hypothesis that Rust's `zopfli` crate only generates DEFLATE blocks with dynamic Huffman codes is false.

@AlexTMjugador Isn't this showing an unrelated feature of `DEFLATE` - the ability to store uncompressed blocks?  IIUC, that's not the same as this request?

**Comment by AlexTMjugador** [comment]
> Isn't this showing an unrelated feature of `DEFLATE` - the ability to store uncompressed blocks? IIUC, that's not the same as this request?

Zopfli is an algorithm that generates DEFLATE streams. Those streams are composed of blocks that may be uncompressed, compressed with fixed Huffman codes, or compressed with dynamic Huffman codes. If I understand this request correctly, the OP wanted a way to instruct encoders (in this case `zlib`, but Zopfli is also a DEFLATE encoder) to generate DEFLATE streams with "fixed Huffman blocks". I have shown that Rust's Zopfli port can indeed generate DEFLATE streams composed of blocks other than dynamically encoded ones, potentially including those with "fixed Huffman blocks", so I'd consider this request resolved, unless there is some sort of bug in the Rust source code that prevents Zopfli from choosing to use fixed Huffman blocks as it is meant to.

In other words: storing uncompressed blocks is indeed a feature of DEFLATE, but it is the encoder who chooses whether an uncompressed, fixed or dynamic block is used, and Rust's Zopfli does indeed not always use dynamic blocks.

**Comment by TPS** [comment]
So, it looks resolved that `oxipng` _can_ do this, @ least internally as part of tests.  Now the question remains whether if it's ever settled on such:
> Is there _any **known**_ example of such a PNG, that isn't better compressed dynamically for an overall smaller size, that's purely due solely to this?

E.g., is this all a waste of code/time, if in practice, its never usable?

**Comment by ace-dent** [comment]
Hi, OP here... :wave:
Thanks to everyone who has contributed to this discussion.

>  Isn't this showing an unrelated feature of DEFLATE - the ability to store uncompressed blocks? IIUC, that's not the same as this request?

@TPS - Yes; correct. You may have stored (uncompressed) or deflate blocks. Deflate blocks may then use dynamic or static (/ 'fixed' ) Huffman tables.

>  I have shown that Rust's Zopfli port can indeed generate DEFLATE streams composed of blocks other than dynamically encoded ones, potentially including those with "fixed Huffman blocks", so I'd consider this request resolved...

@AlexTMjugador - No. With "Potentially", you are making a leap-of-faith from a _stored_ block to _static_ block. The lead author of `zopfli` has said it does not emit fixed Huffman tables. BTW, for low level PNG analysis I find [defdb](https://encode.su/threads/1428-defdb-a-tool-to-dump-the-deflate-stream-from-gz-and-png-files) quite nice. 

>  is this all a waste of code/time, if in practice, it's never usable?

@TPS - For the images I typically compress (e.g. small icons, ~16x16px), this _regularly_ provides a useful optimization. Perhaps ~1 in 50 images benefit, to the extent I anticipate it during [batch](https://github.com/ace-dent/pngslim/blob/feature/RandomTrial/pngslim.cmd#L402-L412) optimization. For compressed images that are ~100 – 200 bytes, it is a significant overhead to store the Huffman codes. Although (AFAIK) Fixed Huffman codes were based on non-image data sets (email text?), they can perform acceptably.

> The theoretical potential of the fixed table is that it isn't stored w/in the image, since it's known anyway & so doesn't take up that many bytes of the image size, right? ...Is there any known example of such a PNG, that isn't better compressed dynamically for an overall smaller size, that's purely due solely to this?

@TPS - Yes. The saving is nearly always due to [not having to store the Huffman codes](https://www.w3.org/Graphics/PNG/RFC-1951#fixed) for small images. I have many real word examples (e.g. [1](https://github.com/ace-dent/8x8.me/blob/main/previews/Rotary.png)). I believe I have once or twice (over 10+ yrs!) found an image, where the static codes were optimal by chance. But for large images, dynamic codes should (nearly) always win.

@andrews05 - This issue is still relevant to the optimization of small images. Should I create a new Issue, or could you re-Open and change the Title of this issue? Ideally the compression engine of oxipng should be capable of producing png files with fixed Huffman code blocks.

**Comment by TPS** [comment]
@ace-dent Thanks very much for answering the above questions/points in such a detailed fashion.  Is it possible to give examples of the images that would benefit from this so they could be used for testing, & even potentially included in this repo?

**Comment by AlexTMjugador** [comment]
> No. With "Potentially", you are making a leap-of-faith from a _stored_ block to _static_ block. The lead author of `zopfli` has said it does not emit fixed Huffman tables.

Well, I didn't say anything wrong. Anyone can check the source I linked to: _it's supposed to_ output fixed blocks too, and I haven't ruled out the hypothesis that there's some oversight preventing Zopfli from doing so, which I'd be happy to fix. But fixing oversights is not a feature request. Also, in computer science, I don't find arguments from authority relevant unless they are accompanied by some sort of explanation.

> BTW, for low level PNG analysis I find [defdb](https://encode.su/threads/1428-defdb-a-tool-to-dump-the-deflate-stream-from-gz-and-png-files) quite nice.

Thanks, that's useful!

**Comment by ace-dent** [comment]
@TPS - Recent examples [here](https://github.com/ace-dent/8x8.me/tree/main/previews) (anything under 100 bytes should definitely be static, optimised with OptiPNG in context of `pngslim`). License is CC0, so I'd be honoured if you use any images. If any of you png ninjas can improve the optimization of these files, please feel free to send me a PR! 🥷

**Comment by AlexTMjugador** [comment]
> sorry if my tone was abrasive. I have not seen the `zoplfi` binary I use, output a static Huffman block. But I have not studied the source code. Thank you for your analysis and work on this issue, I do appreciate it :+1:

No problem, it's always good to get to the truth of things :smile:

In the meantime, I repeated the experiment I described with a small file full of zero bytes (generated by `dd if=/dev/zero bs=1 count=100 of=zeros`, to be exact), and `infgen` showed that Rust's Zopfli does indeed generate a fixed block, which is smaller than a dynamic one for this case:

```
! infgen 3.0 output
!
gzip
!
last
fixed
literal 0
match 99 1
end
!
crc
length
```

I can't rule out that there might be some fine-tuning to do here (for example, Rust's Zopfli avoids computing an expensive, but more accurate cost estimate for fixed blocks when certain conditions apply), but fixed blocks are definitely used. The examples you linked to should be quite useful to improve Zopfli if it turns out that it is estimating fixed block costs incorrectly in some cases.

**Comment by TPS** [comment]
Just a note for the future: [`infgen` just added direct `PNG` support](https://github.com/madler/infgen/commit/6f4eddbccb8957516389c7c87f1e255e93ef2689), so that'll be useful for future technical questions like this.

**Comment by andrews05** [comment]
Nice, that does sound useful.

Just as another note: This discussion has been interesting but it remains the case that oxipng has no control over what block types are output by either zopfli or libdeflate. All evidence indicates these are optimal but even if hypothetically they aren't doing what you want, this is the wrong place to ask for changes - you'll need to direct your requests to the relevant libraries.

**Comment by ace-dent** [comment]
Thanks @andrews05. A fair point.
I'm still not sure whether `zopfli` outputs static blocks but from the [source](https://github.com/ebiggers/libdeflate/blob/8e25ebe07c3514f8379187fefb61ed308557fc11/lib/deflate_compress.c#L1678-L1679) of `libdeflate` it should be capable. The real question is whether that capability is present in oxipng...?

I'd be happy to test but there are no current builds I can run... @TPS / @AlexTMjugador - could you run any of the above files through and see if oxipng outputs a static block? This would lay the issue to rest. 🙏

**Comment by andrews05** [comment]
Well, here's what has been confirmed in this thread thus far:
- Unlike zlib, neither libdeflate nor zopfli expose any ability to control their "strategy". (Ergo it isn't possible for oxipng to do anything about it).
- Both libraries do have the ability to automatically choose from all three block types. (The relevant code in zopfli was linked to earlier).
- Alex has demonstrated this ability in zopfli in his post above where he showed it outputting a "fixed" block type (and earlier a "stored" block type).

It would therefore be logical to conclude that automatic block type selection also happens when these libraries are used within oxipng - there's really no reason to think otherwise.

That said, you should be able to verify this yourself in the oxipng 8.0.0 release build. I'm pretty sure nothing relevant has changed since then.

**Comment by TPS** [comment]
@andrews05 IIUC, @ace-dent's concern about 32-bit must about larger files than these — I can't imagine these tiny things taking much time no matter what they are run through & on.

**Comment by ebiggers** [comment]
Hi, just to clarify libdeflate's behavior:

Whenever it decides it's time to output a block, libdeflate calculates exactly which block type is cheapest of the three available block types in DEFLATE (dynamic, static a.k.a. fixed, and uncompressed a.k.a. stored), and it uses that type.

However, the match/literal sequence that is used for compressed blocks has already been chosen at that point, and libdeflate traditionally hasn't tried to optimize it specifically for static blocks.  Thus, while a static block would be used when cheaper than dynamic, it wouldn't necessarily be the "optimal" static block.

libdeflate v1.19 improved on this by making levels 11-12 find and consider the optimal static block in some cases.  As static blocks are rarely useful for large blocks, for performance reasons this extra step is done only if the block length is <= 1000 bytes (for level 11) or <= 10000 bytes (for level 12).

The exact cutoffs are subject to change, but hopefully this helps a bit for tiny PNGs.

**Comment by TPS** [comment]
@ebiggers Thanks very much 🙇🏾‍♂️

**Comment by andrews05** [comment]
Thanks for the info @ebiggers, I have indeed noticed small improvements on tiny files!
I intend to update oxipng to v1.19 shortly. Sad about the Huffman codes with Explorer though, I've noticed the regressions there too. Hm, I don't suppose that could be enforced in raw deflate only, and not in gzip or zlib mode?
